### PR TITLE
Add error explanations for E0185, E0186, E0202, E0326. Fix E0053.

### PR DESCRIPTION
**Pull request by nham**
Changes:
- adds explanations for E0185, E0186, E0202, E0326
- fixes the explanation for E0053. The previous description was too narrow; there are other error cases.
- changes the error message for E0202 to be specific for associated types, since it seems inherent associated constants are implemented.

Part of #24407

### CODE REVIEW

File: src/librustc_typeck/diagnostics.rs [hunk 1]
```diff
@@ -456,6 +443,48 @@ it has been disabled for now.
...
+trait Foo {
+    fn foo(&self);
+}
+
+struct Bar;
+
+impl Foo for Bar {
+    // error, method `foo` has a `&self` declaration in the trait, but not in
+    // the impl
+    fn foo(&self) {}
```
Comment (alexcrichton) [inline src/librustc_typeck/diagnostics.rs]: Did this mistakenly include `&self`?

**Comment by alexcrichton** [comment]
r=me with just a minor nit, thanks!

File: src/librustc_typeck/diagnostics.rs (hunk 1 above)
Comment (nham) [inline src/librustc_typeck/diagnostics.rs]: Yep. Fixed, thanks.

**Comment by nham** [comment]
@alexcrichton I just realized I missed a test with the E0202 change. I've fixed it now.

**Pull request merged**